class-plugin-autoupdate-filter.php

```diff
@@ -37,6 +39,57 @@ public function init() {
 		add_filter( 'auto_plugin_update_send_email', '__return_true', 11 );
 		add_filter( 'auto_theme_update_send_email', '__return_true', 11 );
 
+		// "Disable all autoupdates" toggle (killswitch)
+		add_filter( 'auto_update_plugin', array( $this, 'auto_update_killswitch' ), PHP_INT_MAX, 2 );
+		add_filter( 'auto_update_core', array( $this, 'auto_update_killswitch' ), PHP_INT_MAX, 2 );
+		add_filter( 'auto_update_theme', array( $this, 'auto_update_killswitch' ), PHP_INT_MAX, 2 );
+		add_action( 'admin_init', array( $this, 'killswitch_engaged_admin_warning' ) );
+
+	}
+
+	/**
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings() {
```

Comment:
Thanks for catching this, as it's clearly not perfectly thought out.

Re: "return;"  My understanding is that using "return" is the accepted way to exit out of a function early, and returning a value is something that's optional. Always open to learning otherwise, though :) 

I think this is closer to what I was trying to do: set the "settings" to have an error value if there's an error. such as:

```PHP
private function get_auto_update_settings() {
  $endpoint_url = 'https://opsoasis.wpspecialprojects.com/wp-json/custom/v1/get_autoupdate_settings/';
  $response     = wp_remote_get( $endpoint_url );
  
  // Check for `WP_Error`
  if ( is_wp_error( $response ) ) {
	  $this->settings = [
		  'error' => $response->get_error_message(),
	  ];
	  return;
  }
  
  // Check for non-200 status code
  $response_code = wp_remote_retrieve_response_code($response);
  if ( 200 !== $response_code ) {
	  $this->settings = [
		  'error' => "HTTP request returned status code {$response_code}.",
	  ];
	  return;
  }
  
  // Check if response body exists and is not empty
  $body = wp_remote_retrieve_body($response);
  if ( empty( $body ) ) {
	  $this->settings = [
		  'error' => 'API returned an empty response body.',
	  ];
	  return;
  }
  
  // Try to decode JSON and check for errors
  $data = json_decode( $body, true );
  if ( json_last_error() !== JSON_ERROR_NONE ) {
	  $this->settings = [
		  'error' => 'Failed to decode JSON: ' . json_last_error_msg(),
	  ];
	  return;
  }
  
  $this->settings = $data;
}
 ```

This way, we can check for an 'error' value in `$this->settings` later, in order to pause all autoupdates if there's been an error getting the settings, such as:

```PHP
public function auto_update_killswitch( $update, $item ) {

  if ( $this->settings['error'] || 'on' === $this->settings['team51_autoupdate_settings_disable_all_toggle'] ) {
	  return false;
  }
  
  return $update;
}
```

Thoughts?

Reply:
> My understanding is that using "return" is the accepted way to exit out of a function early, and returning a value is something that's optional.

Well yes, but also no 😅 Technically, [PHP functions **always** return a value](https://www.php.net/manual/en/functions.returning-values.php). If you use `return;` or simply never return at all, then the return value will be `null`.

Once PHP7 came around, it became possible to type-hint `void` as the function's return value and that would throw an error if your function contained anything other than `return;`. However, [even in this instance](https://stackoverflow.com/a/40960562), you could catch the function's output and it would be `null`.

So the return value is definitely optional since you always have a default of `null` regardless.

-----

That being said, it's typically bad practice to have paths in your code that explicitly return something and others that don't. If you use exclusively `return;` in the code, that's fine (the function's return value is then `void` basically). If you never use `return` at all, that's fine. But when you mix `return;` with `return x;` you start having to do more error-handling than it's worth. 

For example, in your original code, you would have two cases: a) the function returns a string which we should know to interpret as an error, or b) the function returns nothing which implies that it returns `null`. In **both cases**, the programmer picking up the code needs to know something: a) a string implies an error, b) if no string, then null because that's how PHP works (for now), and that means success. 

[PHPMD](https://phpmd.org/) would flag this with an error message like "Not all execution paths return a value". It's fine that none do, or that all do, but mixing them gets iffy. 

-----

I like your current code -- it fixes the mixed paths issue, but I think it's needlessly complicated. The `settings` object now fulfils two separate functions, can have two separate sets of properties, and requires checking which object we're currently working with at every step of the way. 

What about something like this:

```
private function get_auto_update_settings(): stdClass|WP_Error {
    $response = wp_remote_get(
        'https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/'
        array( 'Accept' => 'application/json' )
    );
    if ( is_wp_error( $response ) ) {
        return $response;
    }

    $response_code = wp_remote_retrieve_response_code( $response );
    $response_body = wp_remote_retrieve_body( $response );
 
    // Check that the response code is a 2xx code.
    if ( ! \str_starts_with( (string) $response_code, '2' ) ) {
        $response_message = wp_remote_retrieve_response_message( $response );
        return new WP_Error( $response_code, $response_message, $response_body );
    }

    // Parse the response body as JSON.
    try {
        $decoded_body = json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
    } catch ( JsonException $exception ) {
        return new WP_Error( $exception->getCode(), $exception->getMessage(), $exception );
    }

    return $decoded_body;
}

public function init(): void {
    $this->settings = $this->get_auto_update_settings();
    if ( is_wp_error( $this->settings ) ) {
        // Optionally error log, output as an admin notice, no whatever. 
        error_log( "❌ Plugin Autoupdate Filter API Error: { $this->settings->get_error_code() } { $this->settings->get_error_message() }" ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
        add_action(
            'admin_notices',
            function() {
	        $message = $this->settings->get_error_message();
	        printf( '<div class="notice notice-error"><p>%1$s</p></div>', esc_html( $message ) );
            }
        );

        // Disable all updates.
        $this->settings = (object) array( 'disable_all' => true );
    }
  
    // Continue with plugin initialization.
}
```

That would be the "truly" WordPress way of doing things before PHP even supported exceptions. If you want to use a more modern approach similar to how WooCommerce handles things, then may I propose:

```
private function get_auto_update_settings(): stdClass {
    $response = wp_remote_get(
        'https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/'
        array( 'Accept' => 'application/json' )
    );
    if ( is_wp_error( $response ) ) {
        throw new RuntimeException( $response->get_error_message(), $response->get_error_code() );
    }

    $response_code = wp_remote_retrieve_response_code( $response );
    $response_body = wp_remote_retrieve_body( $response );
 
    // Check that the response code is a 2xx code.
    if ( ! \str_starts_with( (string) $response_code, '2' ) ) {
        $response_message = wp_remote_retrieve_response_message( $response );
        throw new Exception( $response_message, $response_code );
    }

    // Parse the response body as JSON.
    return json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
}

public function init(): void {
    try {
        $this->settings = $this->get_auto_update_settings();
    } catch ( Exception $exception ) {
        // Optionally error log, output as an admin notice, no whatever. 

        // Disable all updates.
        $this->settings = (object) array( 'disable_all' => true );
    }
   
     // Continue with plugin initialization.
}
```

Either of these solutions would simplify your code significantly because:

1. All the error checking and logging would happen in one single place.
2. You can start using the settings property **without** doing error checking throughout your code.

You'll notice that I used a `stdClass` object instead of an array. That was just to keep everything an object but ultimately it doesn't matter. An array might be easier to use since you could `??` any unset values to a default inline.

Reply:
This is great! I've reworked the code to follow this paradigm